# Hackernews Data Analysis Project

In this project we will work with this <a href="https://www.kaggle.com/hacker-news/hacker-news-posts">data set</a>. It consists of 300,000 rows which hold data about posts on the community board of the website Hacker News. The data set we will use here is a smaller subset which contains approximately 20,000 rows. We remove the first row as it contains the column headers.

In [115]:
from csv import reader
import datetime as dt
from datetime import datetime
opened = open("hacker_news.csv")
read_file = reader(opened)
hn = list(read_file)
print(hn[:4],"\n")
header = hn[:1]
hn = hn[1:]
print(header, "\n")
print(hn[:4])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']] 

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']] 

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-

In [116]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = str(row[1])
    title = title.lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(title)
len_a = len(ask_posts)
len_s = len(show_posts)
len_o = len(other_posts)
print(len_a, len_s, len_o)
    
    

1744 1162 17194


In [117]:
total_ask_comments = 0
total_ask_posts = 0
total_show_comments = 0
total_show_posts = 0

for row in ask_posts:
    comments = int(row[4])
    total_ask_comments = total_ask_comments + comments
    total_ask_posts += 1

for row in show_posts:
    comments = int(row[4])    
    total_show_comments = total_show_comments + comments
    total_show_posts += 1
    
avg_ask_comments = round((total_ask_comments / total_ask_posts), 2)
avg_show_comments = round((total_show_comments / total_show_posts), 2)
rel_as = round(avg_ask_comments/avg_show_comments, 2)*100
print(avg_ask_comments)
print(avg_show_comments)
print(rel_as)


14.04
10.32
136.0


The average number of comments on ask posts is 14.04. The average number of comments on show posts is 10.32. Ask posts get 36 % more comments on average.

In [118]:
result_list = []
for row in ask_posts:
    created = row[6]
    created = dt.datetime.strptime(created, "%m/%d/%Y %H:%M")
    comments = int(row[4])    
    item = [created, comments]
    result_list.append(item)
    
counts_by_hour = {}
comments_by_hour = {}

# print(result_list)

for row in result_list:
    date = row[0]
    comment = int(row[1])
    hour = date.hour
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
    elif hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment

print(comments_by_hour, "\n")
print(counts_by_hour)

{0: 447, 1: 683, 2: 1381, 3: 421, 4: 337, 5: 464, 6: 397, 7: 267, 8: 492, 9: 251, 10: 793, 11: 641, 12: 687, 13: 1253, 14: 1416, 15: 4477, 16: 1814, 17: 1146, 18: 1439, 19: 1188, 20: 1722, 21: 1745, 22: 479, 23: 543} 

{0: 55, 1: 60, 2: 58, 3: 54, 4: 47, 5: 46, 6: 44, 7: 34, 8: 48, 9: 45, 10: 59, 11: 58, 12: 73, 13: 85, 14: 107, 15: 116, 16: 108, 17: 100, 18: 109, 19: 110, 20: 80, 21: 109, 22: 71, 23: 68}


In [119]:
avg_by_hour = []
for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

print(avg_by_hour)

[[0, 8.127272727272727], [1, 11.383333333333333], [2, 23.810344827586206], [3, 7.796296296296297], [4, 7.170212765957447], [5, 10.08695652173913], [6, 9.022727272727273], [7, 7.852941176470588], [8, 10.25], [9, 5.5777777777777775], [10, 13.440677966101696], [11, 11.051724137931034], [12, 9.41095890410959], [13, 14.741176470588234], [14, 13.233644859813085], [15, 38.5948275862069], [16, 16.796296296296298], [17, 11.46], [18, 13.20183486238532], [19, 10.8], [20, 21.525], [21, 16.009174311926607], [22, 6.746478873239437], [23, 7.985294117647059]]


For the next step we want to sort this lost to identify the most active hours for comments. 

In [120]:
swap_avg_by_hour = []
for row in avg_by_hour:
    hour = row[0]
    comments = row[1]
    hc_list = [comments, hour]
    swap_avg_by_hour.append(hc_list)
print(swap_avg_by_hour, "\n")
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap, "\n")
print("Top 5 Hours for Ask Posts Comments:\n")

for row in sorted_swap[:5]:
    hour = row[1]
    hour = str(hour)
    hour = hour+":00"
    comments = row[0]
    hour = dt.datetime.strptime(hour, "%H:%M")
    hour = dt.datetime.strftime(hour, "%H:%M")
    print("{0}: {1:.2f} average comments per post.".format(hour, comments))

[[8.127272727272727, 0], [11.383333333333333, 1], [23.810344827586206, 2], [7.796296296296297, 3], [7.170212765957447, 4], [10.08695652173913, 5], [9.022727272727273, 6], [7.852941176470588, 7], [10.25, 8], [5.5777777777777775, 9], [13.440677966101696, 10], [11.051724137931034, 11], [9.41095890410959, 12], [14.741176470588234, 13], [13.233644859813085, 14], [38.5948275862069, 15], [16.796296296296298, 16], [11.46, 17], [13.20183486238532, 18], [10.8, 19], [21.525, 20], [16.009174311926607, 21], [6.746478873239437, 22], [7.985294117647059, 23]] 

[[38.5948275862069, 15], [23.810344827586206, 2], [21.525, 20], [16.796296296296298, 16], [16.009174311926607, 21], [14.741176470588234, 13], [13.440677966101696, 10], [13.233644859813085, 14], [13.20183486238532, 18], [11.46, 17], [11.383333333333333, 1], [11.051724137931034, 11], [10.8, 19], [10.25, 8], [10.08695652173913, 5], [9.41095890410959, 12], [9.022727272727273, 6], [8.127272727272727, 0], [7.985294117647059, 23], [7.852941176470588, 

We can summarize our findings like this: The time between 15:00-17:00, 2:00-3:00 and 20:00-22:00 are the most popular for comments.